src/form-validation.ts

```diff
@@ -0,0 +1,108 @@
+import { ValidationEngine } from './validation-engine';
+import {
+  ValidationSchema,
+  RecordValidationSchema,
+  FieldsValidationSchema,
+  FieldValidation,
+  ValidationResult,
+  FormValidationResult,
+} from './model';
+import { isFunction } from './helper';
+
+export class FormValidation {
+  private validationEngine: ValidationEngine;
+
+  constructor(validationSchema: ValidationSchema) {
+    this.validationEngine = new ValidationEngine();
+    this.setupValidationSchema(validationSchema);
+  }
+
+  // TODO: review parameters order, baseFormValidation validateField has fieldId, value, values
+  // the engine validateField has a different order, IMHO it should be always: fieldId, value, values
+  // since values is options.
+  public validateField(
+    fieldId: string,
+    value: any,
+    values?: any
+  ): Promise<ValidationResult> {
+    return this.validationEngine.validateField(values, fieldId, value);
+  }
+
+  public validateForm(values: any): Promise<FormValidationResult> {
+    return this.validationEngine.validateForm(values);
+  }
+
+  private setupValidationSchema(validationSchema: ValidationSchema) {
+    if (validationSchema && typeof validationSchema === 'object') {
+      const { record, fields } = validationSchema;
+
+      if (record && record instanceof Array) {
+        this.addRecordValidations(record);
+      }
+      if (fields && typeof fields === 'object') {
+        this.addAllFieldsValidations(fields);
+      }
+    } else {
+      console.error('ValidationSchema must be a valid object');
+    }
+  }
+
+  private addRecordValidations(
+    recordValidationSchemaCollection: RecordValidationSchema[]
+  ) {
+    recordValidationSchemaCollection.forEach(
+      (recordValidation: RecordValidationSchema) => {
+        this.addRecordValidation(recordValidation);
+      }
+    );
+  }
+
+  private addRecordValidation(recordValidation: RecordValidationSchema) {
+    if (isFunction(recordValidation)) {
+      this.validationEngine.addRecordValidation(recordValidation);
+    } else {
+      this.validationEngine.addRecordValidation(
```

Comment:
same here

Reply:
Here maybe it could make sense, we got two flavours here:
   - A record validation that can be a function.
   - A record validation that can be a given structure.

One thing that we could do is the other way around, addRecordValidation builds up the RecordValidation structure if needed and then in the engine we just pass the record validation. What do you think?

I think we may go for this second option

Reply:
Done you can see this update in the refactor branch

Reply:
So, you mean to convert to structure on top of validation engine, and propagate it to rest of code. So, we have not to worry about it more, isn't it?

I agree